<a href="https://colab.research.google.com/github/xiborxu/IST462xibo/blob/main/FinalExam_IST462_Fall2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Final exam - IST 462 - Notebook

Submitted by:

Date:

Using the SY36 bus route datasets and the 2019 daily weather conditions for Syracuse data perform the data analysis tasks listed in this document. Perform as many tasks and subtasks as needed to get up to 100 points. If the tasks you complete exceed 100 points, and they are all done correctly, your maximum grade will be capped at 100 points.

Each task has a point value for completing it. There are sub-tasks associated with each task that provide additional points. You can complete a sub-task ONLY if you have completed the main task to which it is associated.

To get all the points for a task or sub-task, in addition to performing the required data analysis mentioned in the task/subtask you must also provide clear documentation of the steps you used in the analysis and in all places possible meaningful graphs that show your results.

**Final exam submission:** Submit your exam work in Notebook format (.ipynb file) ***AND*** in PDF format.

#Import packages

You, of course, can and should add more dependent on what you use in your analysis

In [52]:
%matplotlib inline

import pandas as pd
import numpy as np
import requests
from zipfile import ZipFile
import matplotlib.pyplot as plt
np.set_printoptions(precision=4)
pd.options.display.max_rows = 20

# Loading data sets

In [53]:
# connect colab to google drive

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### 2019 Syracuse weather data

In [54]:
import pandas as pd

file = '/content/drive/MyDrive/Final_syracuse_2019_weather.csv'
weather2019 = pd.read_csv(file) #read in the csv file using pandas
weather2019.head()

STATION                                           NAME      DATE  \
0  USW00014771  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  1/1/2019   
1  USW00014771  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  1/2/2019   
2  USW00014771  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  1/3/2019   
3  USW00014771  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  1/4/2019   
4  USW00014771  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  1/5/2019   

    AWND  PRCP  SNOW  TAVG  TMAX  TMIN  
0  15.66  0.02   0.0    40    53    21  
1   5.14  0.00   0.0    24    31    18  
2  10.74  0.09   0.3    33    37    30  
3   4.70  0.00   0.0    36    49    25  
4   5.59  0.00   0.0    33    44    25

In [55]:
weather2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   STATION  365 non-null    object 
 1   NAME     365 non-null    object 
 2   DATE     365 non-null    object 
 3   AWND     365 non-null    float64
 4   PRCP     365 non-null    float64
 5   SNOW     365 non-null    float64
 6   TAVG     365 non-null    int64  
 7   TMAX     365 non-null    int64  
 8   TMIN     365 non-null    int64  
dtypes: float64(3), int64(3), object(3)
memory usage: 25.8+ KB


## Bus ridership data sets

In [56]:
#Dataframe with bus activity data for 2019 for route SY36
file = '/content/drive/MyDrive/Final_Preprocessed_SY36_2019.csv'
bus_data2019 = pd.read_csv(file) #read in the csv file using pandas
bus_data2019.head()

<ipython-input-56-f3ba1b0c73d7>:3: DtypeWarning: Columns (23) have mixed types. Specify dtype option on import or set low_memory=False.
  bus_data2019 = pd.read_csv(file) #read in the csv file using pandas


SERIAL_NUMBER  SCHEDULE_ID    SCHEDULE_NAME                 SURVEY_DATE  \
0        2604260          295  Sep18 (Holiday)  2019-01-01 00:00:00.000000   
1        2604260          295  Sep18 (Holiday)  2019-01-01 00:00:00.000000   
2        2604260          295  Sep18 (Holiday)  2019-01-01 00:00:00.000000   
3        2604260          295  Sep18 (Holiday)  2019-01-01 00:00:00.000000   
4        2604260          295  Sep18 (Holiday)  2019-01-01 00:00:00.000000   

   PATTERN_ID  ROUTE_NUMBER ROUTE_NAME DIRECTION_NAME  \
0   180900591           371       SY36       FROM HUB   
1   180900591           371       SY36       FROM HUB   
2   180900591           371       SY36       FROM HUB   
3   180900591           371       SY36       FROM HUB   
4   180900591           371       SY36       FROM HUB   

                         BRANCH             TRIP_START_TIME  ...  \
0  [Sy36]Outbound 136 no plazas  2019-01-01 07:40:00.000000  ...   
1  [Sy36]Outbound 136 no plazas  2019-01-01 07:40:00.000000  ...   
2  [Sy36]Outbound 136 no plazas  2019-01-01 07:40:00.000000  ...   
3  [Sy36]Outbound 136 no plazas  2019-01-01 07:40:00.000000  ...   
4  [Sy36]Outbound 136 no plazas  2019-01-01 07:40:00.000000  ...   

  PASSENGERS_ON PASSENGERS_OFF  PASSENGERS_IN  WHEELCHAIRS  TIMEPOINT_MILES  \
0             8              0              8            0            0.413   
1             0              0              8            0              NaN   
2             0              0              8            0            0.716   
3             0              0              8            0              NaN   
4             0              0              8            0              NaN   

   FIRST_LAST_STOP MODIFIED    UNIQUE_ID   stop_lat   stop_lon  
0                1        1  37100000002  43.043656 -76.150963  
1                2        0  37100000003  43.044280 -76.147495  
2                2        1  37100000005  43.045336 -76.147419  
3                2        0  37100000006  43.047959 -76.147440  
4                2        0  37100000007  43.049554 -76.148697  

[5 rows x 47 columns]

In [57]:
bus_data2019.shape

(980551, 47)

In [58]:
#Drop some columns that won't be needed
bus_data2019.drop(['SURVEY_DATE','VEHICLE_DESCRIPTION','GARAGE_NAME','DIVISION_NAME','COMMENTS','WHEELCHAIRS','MODIFIED'],axis=1,inplace=True)

In [59]:
bus_data2019.head()

SERIAL_NUMBER  SCHEDULE_ID    SCHEDULE_NAME  PATTERN_ID  ROUTE_NUMBER  \
0        2604260          295  Sep18 (Holiday)   180900591           371   
1        2604260          295  Sep18 (Holiday)   180900591           371   
2        2604260          295  Sep18 (Holiday)   180900591           371   
3        2604260          295  Sep18 (Holiday)   180900591           371   
4        2604260          295  Sep18 (Holiday)   180900591           371   

  ROUTE_NAME DIRECTION_NAME                        BRANCH  \
0       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   
1       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   
2       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   
3       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   
4       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   

              TRIP_START_TIME TIME_PERIOD  ... DWELL_TIME  \
0  2019-01-01 07:40:00.000000     AM Peak  ...        NaN   
1  2019-01-01 07:40:00.000000     AM Peak  ...        0.0   
2  2019-01-01 07:40:00.000000     AM Peak  ...        0.0   
3  2019-01-01 07:40:00.000000     AM Peak  ...        0.0   
4  2019-01-01 07:40:00.000000     AM Peak  ...        0.0   

   RUNNING_TIME_ACTUAL  PASSENGERS_ON  PASSENGERS_OFF  PASSENGERS_IN  \
0                5.483              8               0              8   
1                  NaN              0               0              8   
2                2.550              0               0              8   
3                  NaN              0               0              8   
4                  NaN              0               0              8   

  TIMEPOINT_MILES  FIRST_LAST_STOP    UNIQUE_ID   stop_lat   stop_lon  
0           0.413                1  37100000002  43.043656 -76.150963  
1             NaN                2  37100000003  43.044280 -76.147495  
2           0.716                2  37100000005  43.045336 -76.147419  
3             NaN                2  37100000006  43.047959 -76.147440  
4             NaN                2  37100000007  43.049554 -76.148697  

[5 rows x 40 columns]

In [60]:
bus_data2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 980551 entries, 0 to 980550
Data columns (total 40 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   SERIAL_NUMBER        980551 non-null  int64  
 1   SCHEDULE_ID          980551 non-null  int64  
 2   SCHEDULE_NAME        980551 non-null  object 
 3   PATTERN_ID           980551 non-null  int64  
 4   ROUTE_NUMBER         980551 non-null  int64  
 5   ROUTE_NAME           980551 non-null  object 
 6   DIRECTION_NAME       980551 non-null  object 
 7   BRANCH               980551 non-null  object 
 8   TRIP_START_TIME      980551 non-null  object 
 9   TIME_PERIOD          980551 non-null  object 
 10  SERVICE_PERIOD       980551 non-null  object 
 11  TRIP_NUMBER          980551 non-null  int64  
 12  TRIP_KEY             980551 non-null  int64  
 13  BLOCK_NUMBER         980551 non-null  int64  
 14  BLOCK_KEY            980551 non-null  int64  
 15  BLOCK_NAME       

In [61]:
#Dataframe with bus activity data for 2020 for route SY36
file = '/content/drive/MyDrive//Preprocessed_SY36_2020.csv'
bus_data2020 = pd.read_csv(file) #read in the csv file using pandas
bus_data2020.head()

SERIAL_NUMBER  SCHEDULE_ID    SCHEDULE_NAME                 SURVEY_DATE  \
0        3286134          314  Dec19 (Holiday)  2020-01-01 00:00:00.000000   
1        3286134          314  Dec19 (Holiday)  2020-01-01 00:00:00.000000   
2        3286134          314  Dec19 (Holiday)  2020-01-01 00:00:00.000000   
3        3286134          314  Dec19 (Holiday)  2020-01-01 00:00:00.000000   
4        3286134          314  Dec19 (Holiday)  2020-01-01 00:00:00.000000   

   PATTERN_ID  ROUTE_NUMBER ROUTE_NAME DIRECTION_NAME  \
0   191200591           371       SY36       FROM HUB   
1   191200591           371       SY36       FROM HUB   
2   191200591           371       SY36       FROM HUB   
3   191200591           371       SY36       FROM HUB   
4   191200591           371       SY36       FROM HUB   

                         BRANCH             TRIP_START_TIME  ...  \
0  [sy36]Outbound 136 no plazas  2020-01-01 07:40:00.000000  ...   
1  [sy36]Outbound 136 no plazas  2020-01-01 07:40:00.000000  ...   
2  [sy36]Outbound 136 no plazas  2020-01-01 07:40:00.000000  ...   
3  [sy36]Outbound 136 no plazas  2020-01-01 07:40:00.000000  ...   
4  [sy36]Outbound 136 no plazas  2020-01-01 07:40:00.000000  ...   

  PASSENGERS_ON PASSENGERS_OFF  PASSENGERS_IN  WHEELCHAIRS  TIMEPOINT_MILES  \
0             9              0              9            0            0.417   
1             0              0              9            0              NaN   
2             0              0              9            0            0.705   
3             0              0              9            0              NaN   
4             1              0             10            0              NaN   

   FIRST_LAST_STOP MODIFIED    UNIQUE_ID   stop_lat   stop_lon  
0                1        1  37100000002  43.043656 -76.150963  
1                2        0  37100000003  43.044280 -76.147495  
2                2        1  37100000005  43.045336 -76.147419  
3                2        0  37100000006  43.047959 -76.147440  
4                2        0  37100000007  43.049554 -76.148697  

[5 rows x 47 columns]

In [62]:
#Drop some columns that won't be needed
bus_data2020.drop(['SURVEY_DATE','VEHICLE_DESCRIPTION','GARAGE_NAME','DIVISION_NAME','COMMENTS','WHEELCHAIRS','MODIFIED'],axis=1,inplace=True)

In [63]:
bus_data2020.head()

SERIAL_NUMBER  SCHEDULE_ID    SCHEDULE_NAME  PATTERN_ID  ROUTE_NUMBER  \
0        3286134          314  Dec19 (Holiday)   191200591           371   
1        3286134          314  Dec19 (Holiday)   191200591           371   
2        3286134          314  Dec19 (Holiday)   191200591           371   
3        3286134          314  Dec19 (Holiday)   191200591           371   
4        3286134          314  Dec19 (Holiday)   191200591           371   

  ROUTE_NAME DIRECTION_NAME                        BRANCH  \
0       SY36       FROM HUB  [sy36]Outbound 136 no plazas   
1       SY36       FROM HUB  [sy36]Outbound 136 no plazas   
2       SY36       FROM HUB  [sy36]Outbound 136 no plazas   
3       SY36       FROM HUB  [sy36]Outbound 136 no plazas   
4       SY36       FROM HUB  [sy36]Outbound 136 no plazas   

              TRIP_START_TIME TIME_PERIOD  ... DWELL_TIME  \
0  2020-01-01 07:40:00.000000     AM Peak  ...        NaN   
1  2020-01-01 07:40:00.000000     AM Peak  ...       0.00   
2  2020-01-01 07:40:00.000000     AM Peak  ...       0.00   
3  2020-01-01 07:40:00.000000     AM Peak  ...       0.00   
4  2020-01-01 07:40:00.000000     AM Peak  ...       0.12   

   RUNNING_TIME_ACTUAL  PASSENGERS_ON  PASSENGERS_OFF  PASSENGERS_IN  \
0                5.533              9               0              9   
1                  NaN              0               0              9   
2                4.200              0               0              9   
3                  NaN              0               0              9   
4                  NaN              1               0             10   

  TIMEPOINT_MILES  FIRST_LAST_STOP    UNIQUE_ID   stop_lat   stop_lon  
0           0.417                1  37100000002  43.043656 -76.150963  
1             NaN                2  37100000003  43.044280 -76.147495  
2           0.705                2  37100000005  43.045336 -76.147419  
3             NaN                2  37100000006  43.047959 -76.147440  
4             NaN                2  37100000007  43.049554 -76.148697  

[5 rows x 40 columns]

In [64]:
bus_data2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 962566 entries, 0 to 962565
Data columns (total 40 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   SERIAL_NUMBER        962566 non-null  int64  
 1   SCHEDULE_ID          962566 non-null  int64  
 2   SCHEDULE_NAME        962566 non-null  object 
 3   PATTERN_ID           962566 non-null  int64  
 4   ROUTE_NUMBER         962566 non-null  int64  
 5   ROUTE_NAME           962566 non-null  object 
 6   DIRECTION_NAME       962566 non-null  object 
 7   BRANCH               962566 non-null  object 
 8   TRIP_START_TIME      962566 non-null  object 
 9   TIME_PERIOD          962566 non-null  object 
 10  SERVICE_PERIOD       962566 non-null  object 
 11  TRIP_NUMBER          962566 non-null  int64  
 12  TRIP_KEY             962566 non-null  int64  
 13  BLOCK_NUMBER         962566 non-null  int64  
 14  BLOCK_KEY            962566 non-null  int64  
 15  BLOCK_NAME       

# Exam tasks

**Task 1 (30 points):** For the year of 2019 determine the number of passengers that board the bus (PASSENGERS_ON) at a particular STOP_ID per day. Use this data to understand how the changes in weather affect the ridership at your selected Bus Stop. Select a bus stop with a daily annual average of at least 5 passengers using it (This means that for any service day of the year, on average, at least 5 passengers boarded the bus from that bus stop).

In [65]:
bus_data2019.head()

SERIAL_NUMBER  SCHEDULE_ID    SCHEDULE_NAME  PATTERN_ID  ROUTE_NUMBER  \
0        2604260          295  Sep18 (Holiday)   180900591           371   
1        2604260          295  Sep18 (Holiday)   180900591           371   
2        2604260          295  Sep18 (Holiday)   180900591           371   
3        2604260          295  Sep18 (Holiday)   180900591           371   
4        2604260          295  Sep18 (Holiday)   180900591           371   

  ROUTE_NAME DIRECTION_NAME                        BRANCH  \
0       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   
1       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   
2       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   
3       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   
4       SY36       FROM HUB  [Sy36]Outbound 136 no plazas   

              TRIP_START_TIME TIME_PERIOD  ... DWELL_TIME  \
0  2019-01-01 07:40:00.000000     AM Peak  ...        NaN   
1  2019-01-01 07:40:00.000000     AM Peak  ...        0.0   
2  2019-01-01 07:40:00.000000     AM Peak  ...        0.0   
3  2019-01-01 07:40:00.000000     AM Peak  ...        0.0   
4  2019-01-01 07:40:00.000000     AM Peak  ...        0.0   

   RUNNING_TIME_ACTUAL  PASSENGERS_ON  PASSENGERS_OFF  PASSENGERS_IN  \
0                5.483              8               0              8   
1                  NaN              0               0              8   
2                2.550              0               0              8   
3                  NaN              0               0              8   
4                  NaN              0               0              8   

  TIMEPOINT_MILES  FIRST_LAST_STOP    UNIQUE_ID   stop_lat   stop_lon  
0           0.413                1  37100000002  43.043656 -76.150963  
1             NaN                2  37100000003  43.044280 -76.147495  
2           0.716                2  37100000005  43.045336 -76.147419  
3             NaN                2  37100000006  43.047959 -76.147440  
4             NaN                2  37100000007  43.049554 -76.148697  

[5 rows x 40 columns]

In [66]:
bus_data2019['STOP_ID'].value_counts()

 3761     12115
 1114     11880
 17661    10824
 14616     9102
 14615     8862
          ...  
 17644      860
 17640      467
-49         292
 17659      233
 17641       99
Name: STOP_ID, Length: 162, dtype: int64

In [67]:
bus_data2019['TRIP_START_TIME'] = pd.to_datetime(bus_data2019['TRIP_START_TIME'])
bus_data2019['DATE'] = bus_data2019['TRIP_START_TIME'].dt.strftime('%m/%d/%Y')

In [68]:
mask = bus_data2019['TRIP_START_TIME'].dt.year != 2019
bus_data2019 = bus_data2019[~mask]

In [69]:
stop_info = bus_data2019[bus_data2019['STOP_ID'] == 3761]
daily_passenger = stop_info.groupby('DATE')['PASSENGERS_ON'].sum().reset_index()
daily_passenger

DATE  PASSENGERS_ON
0    01/01/2019             42
1    01/02/2019             92
2    01/03/2019             93
3    01/04/2019            101
4    01/05/2019             78
..          ...            ...
360  12/27/2019             99
361  12/28/2019             72
362  12/29/2019             68
363  12/30/2019             73
364  12/31/2019            138

[365 rows x 2 columns]

***Subtask 1.1 (+10 points):*** Group the activity at the selected bus stop per month and compare against the average temperature for that month.

In [70]:
weather2019['DATE'] = pd.to_datetime(weather2019['DATE'])

# Now, you can format the 'DATE' column as desired
weather2019['DATE'] = weather2019['DATE'].dt.strftime('%m/%d/%Y')

In [71]:
merged_df = pd.merge(daily_passenger,weather2019, on='DATE', how='inner')
merged_df

DATE  PASSENGERS_ON      STATION  \
0    01/01/2019             42  USW00014771   
1    01/02/2019             92  USW00014771   
2    01/03/2019             93  USW00014771   
3    01/04/2019            101  USW00014771   
4    01/05/2019             78  USW00014771   
..          ...            ...          ...   
360  12/27/2019             99  USW00014771   
361  12/28/2019             72  USW00014771   
362  12/29/2019             68  USW00014771   
363  12/30/2019             73  USW00014771   
364  12/31/2019            138  USW00014771   

                                              NAME   AWND  PRCP  SNOW  TAVG  \
0    SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  15.66  0.02   0.0    40   
1    SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US   5.14  0.00   0.0    24   
2    SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  10.74  0.09   0.3    33   
3    SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US   4.70  0.00   0.0    36   
4    SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US   5.59  0.00   0.0    33   
..                                             ...    ...   ...   ...   ...   
360  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  10.74  0.00   0.0    47   
361  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US   6.71  0.00   0.0    39   
362  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US   7.83  0.29   0.0    32   
363  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US  12.30  0.75   0.0    37   
364  SYRACUSE HANCOCK INTERNATIONAL AIRPORT, NY US   8.72  0.15   0.4    37   

     TMAX  TMIN  
0      53    21  
1      31    18  
2      37    30  
3      49    25  
4      44    25  
..    ...   ...  
360    55    39  
361    42    27  
362    37    25  
363    41    35  
364    43    32  

[365 rows x 10 columns]

In [72]:
merged_df['DATE'] = pd.to_datetime(merged_df['DATE'])

# Then create a 'MONTH' column to use as the index for the pivot table
merged_df['MONTH'] = merged_df['DATE'].dt.strftime('%m')

# Group by month and calculate the sum of 'PASSENGERS_ON' and the mean of 'TAVG'
monthly_activity = merged_df.groupby(merged_df['DATE'].dt.to_period('M')).agg({
    'PASSENGERS_ON': 'sum',  # Sum the passengers per month
    'TAVG': 'mean'  # Average temperature for the month
}).reset_index()

monthly_activity

DATE  PASSENGERS_ON       TAVG
0   2019-01           2200  21.677419
1   2019-02           2109  26.071429
2   2019-03           2384  32.483871
3   2019-04           2315  47.866667
4   2019-05           2521  56.838710
5   2019-06           2257  66.633333
6   2019-07           2469  74.741935
7   2019-08           2600  69.548387
8   2019-09           2663  64.033333
9   2019-10           2760  53.451613
10  2019-11           2528  36.466667
11  2019-12           2685  30.709677

***Subtask 1.2 (+10 points):*** Compare the activity between 2 or more bus stops over each month of the year.

In [73]:
bus_data2019['DATE'] = pd.to_datetime(bus_data2019['DATE'])
bus_data2019['MONTH'] = bus_data2019['DATE'].dt.strftime('%m')

# Filtering the data for the bus stops 3761, 1114, and 17661
filtered_data = bus_data2019[bus_data2019['STOP_ID'].isin([3761, 1114, 17661])]

# Grouping the data by 'MONTH' and 'STOP_ID' and calculating the sum for 'PASSENGERS_ON', 'PASSENGERS_OFF', and 'PASSENGERS_IN'
activity_comparison = filtered_data.groupby(['MONTH', 'STOP_ID']).agg({
    'PASSENGERS_ON': 'sum',
    'PASSENGERS_OFF': 'sum',
    'PASSENGERS_IN': 'sum'
}).reset_index()

activity_comparison

MONTH  STOP_ID  PASSENGERS_ON  PASSENGERS_OFF  PASSENGERS_IN
0     01     1114            603             312           5821
1     01     3761           2200            1565           2865
2     01    17661           4263            3276           4199
3     02     1114            590             289           5555
4     02     3761           2109            1496           2731
..   ...      ...            ...             ...            ...
31    11     3761           2528            1864           3581
32    11    17661           5146            3886           5098
33    12     1114            781             400           7513
34    12     3761           2685            2097           3506
35    12    17661           5264            4039           5211

[36 rows x 5 columns]

***Subtask 1.3 (+10 points):*** Determine the 5 bus stops that provide the highest average number of daily passengers during the year.

In [80]:
daily_passengers = bus_data2019.groupby(['STOP_ID', 'DATE'])['PASSENGERS_IN'].sum().reset_index()
# Group the data again by 'STOP_ID' and calculate the average daily passengers for each stop
average_daily_passengers = daily_passengers.groupby('STOP_ID')['PASSENGERS_IN'].mean().reset_index()
# Sort the stops by average daily passengers in descending order and select the top 5
top_5_stops = average_daily_passengers.sort_values(by='PASSENGERS_IN', ascending=False).head(5)

# Print the top 5 stops
print(top_5_stops)

     STOP_ID  PASSENGERS_IN
132     8694     268.575342
119     7228     268.372603
124     7870     266.646575
116     7191     262.282192
122     7642     261.805479




```
# This is formatted as code
```

**Task 2 (30 points):** For the years of 2019 and 2020, determine the number of passengers that board the bus (PASSENGERS_ON) at a particular STOP_ID per week. Select a bus stop where you would expect a high number of users (i.e. Near a shopping mall, school, hospital, etc) and provide proof of how you found this via bus map, etc. Compare the ridership activity between the two years and state a hypothesis as to why changes could have taken place. Provide evidence to support that your hypothesis is likely true. The evidence could be in the form or newspaper articles, government declarations, etc.

In [82]:
bus_data2020['TRIP_START_TIME'] = pd.to_datetime(bus_data2020['TRIP_START_TIME'])
bus_data2020['DATE'] = bus_data2020['TRIP_START_TIME'].dt.strftime('%m/%d/%Y')

In [83]:
data_2019 = bus_data2019[bus_data2019['STOP_ID'] == 3761]
data_2020 = bus_data2020[bus_data2020['STOP_ID'] == 3761]

# Group data by 'DATE' and calculate weekly total passengers for both years
data_2019['DATE'] = pd.to_datetime(data_2019['DATE'])
data_2020['DATE'] = pd.to_datetime(data_2020['DATE'])

# Group data by week and sum the passengers on each week
weekly_ridership_2019 = data_2019.groupby(data_2019['DATE'].dt.strftime('%U')).agg({'PASSENGERS_ON': 'sum'}).reset_index()
weekly_ridership_2020 = data_2020.groupby(data_2020['DATE'].dt.strftime('%U')).agg({'PASSENGERS_ON': 'sum'}).reset_index()
print(weekly_ridership_2019)
print(weekly_ridership_2020)

   DATE  PASSENGERS_ON
0    00            406
1    01            536
2    02            531
3    03            466
4    04            529
..  ...            ...
48   48            617
49   49            685
50   50            554
51   51            550
52   52            279

[53 rows x 2 columns]
   DATE  PASSENGERS_ON
0    00            356
1    01            524
2    02            483
3    03            610
4    04            564
..  ...            ...
48   48            646
49   49            664
50   50            582
51   51            519
52   52            448

[53 rows x 2 columns]


<ipython-input-83-c132d4b11697>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_2019['DATE'] = pd.to_datetime(data_2019['DATE'])
<ipython-input-83-c132d4b11697>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_2020['DATE'] = pd.to_datetime(data_2020['DATE'])


***Subtast 2.1 (+10 points):*** Determine the 10 bus stops that provide the highest average number of weekly passengers for each year. Provide a brief comment on any differences that you find interesting.

**Task 3 (20 points):** For both the years 2019 and 2020, compute the distance traveled by the buses serving the route per day. Use the data in the SEGMENT_MILES column for this purpose.

***Subtask 3.1 (+10 points):*** Break down the miles travelled by the buses per month. Assuming an efficiency of 7 miles per gallon and a price of US $3.60 per gallow, compute how much it costs to run the buses per month if they use diesel fuel.

***Subtast 3.2 (+10 points):*** Determine if there are any anomalies (outliers) in the values of the daily distance travelled by the bus each year. Give a possible hypothesis for why they may have occurred and provide evicence to support your hypothesis such as an article or government declaration, etc.

**Task 4 (20 points):** For both years 2019 and 2020, study the differences between the time of arrival (TIME_ACTUAL_ARRIVE) and the scheduled bus time (TIME_SCHEDULED) to determine the hours of the day, or the weeks of the year (or some other time period) where significant deviations between the arrival time and the scheduled time appear. Deviations of more than 10 minutes can be considered significant but you can change this limit if it makes sense for your analysis.

***Subtask 4.1 (+10 points):*** For 2019 only investigate if there is any relationship between weather conditions (termperature, wind, snow) with the deviations between time of arrival and scheduled bust time. Justify your analysis with graphs and documentation.

***Subtask 4.2 (+10 points):*** Investigate if the VEHICLE_NUMBER or OPERATOR_ID have an influence in the deviations between time of arrival and scheduled bus time. Justify your analysis with graphs and documentation.